clone/logdb/database.go

```diff
@@ -40,18 +46,78 @@ func NewDatabase(connString string) (*Database, error) {
 	db := &Database{
 		db: dbConn,
 	}
-	return db, db.Init()
+	return db, db.init()
+}
+
+// NewDatabase creates a Database using the given database connection.
+func NewDatabaseDirect(db *sql.DB) (*Database, error) {
+	ret := &Database{
+		db: db,
+	}
+	return ret, ret.init()
 }
 
-// Init creates the database tables if needed.
-func (d *Database) Init() error {
+func (d *Database) init() error {
 	if _, err := d.db.Exec("CREATE TABLE IF NOT EXISTS leaves (id INTEGER PRIMARY KEY, data BLOB)"); err != nil {
 		return err
 	}
-	// TODO(mhutchinson): Create a table for checkpoints too?
+	if _, err := d.db.Exec("CREATE TABLE IF NOT EXISTS checkpoints (size INTEGER PRIMARY KEY, data BLOB, compactRange BLOB)"); err != nil {
+		return err
+	}
 	return nil
 }
 
+// WriteCheckpoint writes the checkpoint for the given tree size.
+// This should have been verified before writing.
+func (d *Database) WriteCheckpoint(ctx context.Context, size uint64, checkpoint []byte, compactRange [][]byte) error {
+	tx, err := d.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("BeginTx(): %v", err)
+	}
+
+	row := tx.QueryRowContext(ctx, "SELECT size FROM checkpoints ORDER BY size DESC LIMIT 1")
+	var max uint64
+	if err := row.Scan(&max); err != nil {
+		if err != sql.ErrNoRows {
+			tx.Rollback()
+			return fmt.Errorf("Scan(): %v", err)
+		}
+	}
+
+	if size <= max {
+		tx.Rollback()
+		return nil
+	}
+
+	srs := make([]byte, hashSizeBytes*len(compactRange))
+	for i, cr := range compactRange {
+		from := i * hashSizeBytes
+		to := from + hashSizeBytes
+		copy(srs[from:to], cr)
+	}
+	tx.ExecContext(ctx, "INSERT INTO checkpoints (size, data, compactRange) VALUES (?, ?, ?)", size, checkpoint, srs)
+	return tx.Commit()
+}
+
+// GetLatestCheckpoint gets the details of the latest checkpoint.
+func (d *Database) GetLatestCheckpoint(ctx context.Context) (size uint64, checkpoint []byte, compactRange [][]byte, err error) {
+	row := d.db.QueryRowContext(ctx, "SELECT size, data, compactRange FROM checkpoints ORDER BY size DESC LIMIT 1")
+	srs := make([]byte, 0)
+	if err := row.Scan(&size, &checkpoint, &srs); err != nil {
+		if err == sql.ErrNoRows {
+			return 0, nil, nil, ErrNoDataFound
+		}
+		return 0, nil, nil, fmt.Errorf("Scan(): %v", err)
+	}
+	compactRange = make([][]byte, len(srs)/hashSizeBytes)
+	for i := range compactRange {
+		from := i * hashSizeBytes
+		to := from + hashSizeBytes
+		compactRange[i] = srs[from:to]
+	}
+	return
```

Comment:
Perhaps you could use GOB encoding of the array rather than doing it manually?

Reply:
Now done. Ta.